compiler/symbol.go

```diff
@@ -292,10 +294,39 @@ func (c *compilerContext) parsePragmas(info *functionInfo, f *ssa.Function) {
 				if len(parts) != 3 {
 					continue
 				}
-				c.checkWasmImport(f, comment.Text)
+				if f.Blocks != nil {
+					// Defined functions cannot be exported.
+					c.addError(f.Pos(), "can only use //go:wasmimport on declarations")
+					continue
+				}
+				c.checkWasmImportExport(f, comment.Text)
 				info.exported = true
 				info.wasmModule = parts[1]
 				info.wasmName = parts[2]
+			case "//go:wasmexport":
+				if f.Blocks == nil {
+					c.addError(f.Pos(), "can only use //go:wasmexport on definitions")
+					continue
+				}
+				if len(parts) != 2 {
+					c.addError(f.Pos(), fmt.Sprintf("expected one parameter to //go:wasmimport, not %d", len(parts)-1))
+					continue
+				}
+				name := parts[1]
+				if name == "_start" || name == "_initialize" {
+					c.addError(f.Pos(), fmt.Sprintf("//go:wasmexport does not allow %#v", name))
+					continue
+				}
+				if c.BuildMode != "c-shared" && f.RelString(nil) == "main.main" {
+					c.addError(f.Pos(), fmt.Sprintf("//go:wasmexport does not allow main.main to be exported with -buildmode=%s", c.BuildMode))
```

Comment:
Does big Go blow up if `main.main` is declared with `-buildmode=c-shared`?

Reply:
Yes. This is what the proposal says:

> When the `-buildmode` flag is unset, the `_start` function and any exported functions will be exported to the host. Using `//go:wasmexport` on the `main` function in this mode will result in a compilation error. In this mode, only `_start` will initialize the runtime, and so must be the first export called from the host. Any other exported functions may only be called through calling into host functions that call other exports during the execution of the `_start` function. Once the `_start` function has returned, no other exports may be called on the same instance.

Reply:
I think the proposal allows main.main, and wasmexport on it, when -buildmode=c-shared:

> When using -buildmode=c-shared, the resulting Wasm binary will not contain a _start function, and will only contain the _initialize function and any exported functions. The Go main function will not be exported to the host. The user can choose to export it like any other function using the //go:wasmexport directive. The _initialize function will not automatically call main. The main function will not initialize the runtime.

Reply:
Ah, I missed the ! in the line. My mistake!